src/simulator_core/solver/matrix/matrix.py

```diff
@@ -42,58 +42,48 @@ def add_unknowns(self, number_unknowns: int) -> int:
         if number_unknowns < 1:
             raise ValueError("Number of unknowns should be at least 1.")
         self.num_unknowns += number_unknowns
-        self.sol_new += [1.0] * number_unknowns
-        self.sol_old += [0.0] * number_unknowns
+        self.sol_new = np.concatenate([self.sol_new, np.array([1.0] * number_unknowns)])
+        self.sol_old = np.concatenate([self.sol_old, np.array([0.0] * number_unknowns)])
         return self.num_unknowns - number_unknowns
 
-    def add_equation(self, equation_object: EquationObject) -> None:
-        """Method to add an equation to the matrix.
-
-        Add equation to the matrix it returns a unique id to the equation for easy access.
-        :param EquationObject equation_object: Object containing all information of the equation
-        :return:
-        """
-        row = len(self.rhs)
-        self.rhs.append(0.0)
-        self.mat.append([])
-        self.mat[row] = equation_object.to_list(self.num_unknowns)
-        self.rhs[row] = equation_object.rhs
-
     def solve(self, equations: list[EquationObject], dumb: bool = False) -> list[float]:
-        """Method to solve the system of equation given in the matrix.
+        """Method to solve the system of equation given in the matrix using sparse matrix solver.
 
-        Solves the system of equations and returns the solution. The numpy linalg
-        library solve method is used. For this the matrix is converted to np arrays.
+        :param dumb:
+        :param equations: list with the equations to solve.
         :return: list containing the solution of the system of equations.
         """
-        # TODO add checks if enough equations have been supplied.
-        # TODO check if matrix is solvable.
-        self.rhs = []
-        self.mat = []
-        for equation in equations:
-            self.add_equation(equation)
-        if dumb:
-            self.dump_matrix()
+        if len(equations) > self.num_unknowns:
+            raise RuntimeError(f"Too many equations supplied. Got {len(equations)} equations, "
+                               f"but number of unknowns is {self.num_unknowns}")
+        if len(equations) < self.num_unknowns:
+            raise RuntimeError(f"Not enough equation supplied. Got {len(equations)} equations, "
+                               f"but number of unknowns is {self.num_unknowns}")
         self.sol_old = self.sol_new
-        a = np.array(self.mat)
-        b = np.array(self.rhs)
-        self.sol_new = np.linalg.solve(a, b).tolist()
-        return self.sol_new
+        data = np.concatenate([equation.coefficients for equation in equations])
```

Comment:
I had my doubts about this concatenation but GitHub Copilot chat agrees with this approach as long as the concatenation process is the limiting factor.

Reply:
Can you maybe add a few lines of description to detail what you are doing here? In addition, I would rename data to equation_coefficient_array.

Reply:
done